### ISSUE TITLE
Can't glob files in sub-directories

### ISSUE THREAD

**Issue by chapati23** [timkendrick/recursive-copy#3]
Hi, first of all thanks for this package. It's exactly what I need, small and lightweight :)

I have a directory structure like this:
```
.
..
dist/
src/
— sections/
    — section1.html
    — section2.html
— index.html
— index.css
```

My copy task looks like this:
```
copy('src', 'dist', {
    filter: ['!index.html', 'sections/**/*.html'],
    overwrite: true
})
```

Expected outcome: Copy files section1.html and section2.html to dist/sections/section*.html
Real outcome: Nothing copied.

Either I don't understand the globbing mechanism correctly or something's wrong with the implementation of it.

Feedback appreciated :)

**Comment by timkendrick** [comment]
Thanks for the feedback. I agree that the array glob use case is a bit counterintuitive at the moment, here's a bit more information on how it works:

- By default, all files/folders will be copied
- If you pass an array as the `filter` option, it will only copy files that match *all* the array conditions (rather than *any* of the array of conditions)
- The filter is applied to directory paths as well as file paths – so if a directory doesn't match the filter, none of its children will be copied

In your example, it doesn't copy any files for the following reasons:

- The `index.css` doesn't match the `sections/**/*.html` filter, so that won't be copied
- The `sections` directory itself doesn't match the `sections/**/*.html` filter, so nothing inside it will be copied

In the simple case you've mentioned, you could just pass the string `"!index.html"` as the `filter` option. That will copy the `index.css` and everything in the `sections` directory (and anything else that happens to be in the `src` folder).

I'm guessing you've got other files that you want to exclude in there? If this is the case then let me know and I'll see how I can help.

N.B. I agree that both the "must-match-every-condition" filter behavior and the parent-directory-glob-matching can seem a bit counter-intuitive. I'll have a think about whether I can justify a breaking change (and therefore a new major version) to change this behavior...

**Comment by chapati23** [comment]
Hey, thanks for the explanations.
For the time being I involved to multiple more explicit copy-tasks.

I believe switching the array-logic to _any_ conditions vs _all_ would be more in line with how most packages I know have implemented globbing leading to a more intuitive understanding at first use ;)

**Comment by timkendrick** [comment]
I've been thinking this over and I agree that it's worth releasing a new major version that uses an inclusive "any" filter array (rather than the current exclusive "all" filter array) to make it more intuitive.

I've started the `feature-inclusive-filters` branch to track this change (the fix is at commit 6a6965e). While this combines the filters correctly, it still doesn't solve the second issue (i.e. globs aren't matching because a parent directory doesn't match the glob pattern).

This means the tests are currently failing due to the parent directory paths themselves not matching the filter – at the moment it's a fairly dumb filesystem walker so I'll need to change the way that directories are copied to get round this (i.e. lazily create a new directory only once you know there are files inside it that need copying, or if the source directory is empty, etc etc).

I should be able to spend a bit of time on this at some point in the next few days – will keep you posted if I hit any major setbacks.

**Comment by chapati23** [comment]
Great, thanks for the effort. Would love to jump in and help out but covered head-to-toe in client work atm :(

**Comment by timkendrick** [comment]
No worries, I know the feeling! I should be able to get it done at some point over the next week – bug me if it starts looking like I've forgotten about it...

**Comment by timkendrick** [comment]
Sorry for the delay, I've set aside some time to work on this today and tomorrow.

I've been thinking about this change over the last few days, and there are a couple of subtleties that the implementation will have to cover:

- Negated globs: an array of globs involving at least one negation is a bit more complex than just applying a simple `filters.some()` operation, as you need to acknowledge the order of the globs. For example, most implementations (e.g. [multimatch](https://github.com/sindresorhus/multimatch)) seem to say that a glob array like `['assets/**/*', 'src/**/*', '!src/**/*.html', 'src/index.html']` means "all of the 'assets' directory, and all of the 'src' directory, *except* any 'html' files in the 'src' directory, plus the 'index.html' file in the 'src' directory. In logic terms, that boils down to `(((a || b) && !c) || d)`, so any negated glob is applied to the union of all the conditions prior to the negation. To keep things simple I'll split the filter-matching out into a separate package based on multimatch but with function and regexp support (I proposed a [PR](https://github.com/sindresorhus/multimatch/pull/13) to multimatch a while back, they wanted to keep that glob-only so I'll make a separate package that adds the extra functionality).

- Parent directories: say you have a glob such as `'src/**/*'`, however the 'src' directory is empty. As per standard globbing behavior, this directory will not be copied (unless you add an extra `'src'` glob). This is to be expected. If the 'src' directory is *not* empty, however, any files and/or directories inside the 'src' directory, *will* be copied, seeing as they match the glob. This means that while scanning the source directory, when we reach the 'src' folder, we have to make a decision: should it create the directory or not? It's impossible to tell at this point without checking whether there are files within the directory that match a glob pattern and therefore mean that the directory has to be created retrospectively. This is particularly apparent if the glob has multiple levels – e.g. for the glob `'src/app/templates/*'`, if the 'templates' directory is empty then none of the parent directories should be created (potentially slightly counterintuitive, but this is how all the other libraries seem to do it – plus you can always add a `'src/app/templates'` pattern to force this directory to be created). I reckon that the simplest way to implement this functionality is to recursively walk all the source directory's children using nested `readdir()` calls, build a compete file/directory list, and then apply any filters to this filelist and copy the resulting filtered set of files and directories. I'm aware that this is potentially non-optimal when it comes to performance (you end up walking trees that are going to be filtered out), however it should keep the code much simpler when it comes to determining whether a directory should be created or not, and I'd much rather make a few unnecessary `readdir()` calls than introduce more complex logic to determine whether a parent directory could potentially be matched by a glob.

I think this should cover everything, apologies if this is way too much detail – changing the filter behavior necessitates a new major version so I might as well get any other big changes out of the way while I'm at it. Let me know if there's anything you think I've missed!

**Comment by chapati23** [comment]
yeah, i was gonna say why not just build on top of multimatch ;)
so sounds good, nothing to add here!

**Comment by timkendrick** [comment]
I've pushed a new version that includes these changes (2.0.0-beta). It all seems fine on my machine and on the CI tests – could you run an `npm install recursive-copy@2.0.0-beta` on your project and check that it works for you?

Let me know if that's all good so I can publish the stable release to npm.

**Comment by chapati23** [comment]
works like a charm! just tested against v1 and v2 and it now results in the expected outcome (see above).

thanks for your work :)

**Comment by timkendrick** [comment]
No problem! I've published version 2.0.0 to npm now.

(This is a new major release, so you'll need to run `npm install --save recursive-copy@2` to upgrade any existing projects that depend on the old version).

Good luck!
